include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2261,147 +2261,24 @@ namespace internal
-                                             fe_eval);
-        }
-      else if (element_type == ElementType::tensor_raviart_thomas)
-        {
-          FEEvaluationImpl<
-            ElementType::tensor_raviart_thomas,
-            dim,
-            (fe_degree == -1) ? 1 : fe_degree,
-            (n_q_points_1d < 1) ? 1 : n_q_points_1d,
-            Number>::template evaluate_or_integrate<false>(evaluation_flag,
-                                                           const_cast<Number *>(
-                                                             values_dofs),
-                                                           fe_eval);
-        }
-      else
-        {
-          FEEvaluationImpl<ElementType::tensor_general,
-                           dim,
-                           fe_degree,
-                           n_q_points_1d,
-                           Number>::evaluate(n_components,
-                                             evaluation_flag,
-                                             values_dofs,
-                                             fe_eval);
-        }
-
-      return false;
-    }
-  };
-
-
-
-  /**
-   * This class chooses an appropriate evaluation strategy based on the
-   * template parameters and the shape_info variable which contains runtime
-   * parameters for the strategy underlying FEEvaluation::integrate(), i.e.
-   * this calls internal::FEEvaluationImpl::integrate(),
-   * internal::FEEvaluationImplCollocation::integrate() or
-   * internal::FEEvaluationImplTransformToCollocation::integrate() with
-   * appropriate template parameters. In case the template parameters
-   * fe_degree and n_q_points_1d contain valid information (i.e. fe_degree>-1
-   * and n_q_points_1d>0), we simply pass these values to the respective
-   * template specializations.  Otherwise, we perform a runtime matching of
-   * the runtime parameters to find the correct specialization. This matching
-   * currently supports $0\leq fe\_degree \leq 9$ and $degree+1\leq
-   * n\_q\_points\_1d\leq fe\_degree+2$.
-   */
-  template <int dim, typename Number>
-  struct FEEvaluationImplIntegrateSelector
-  {
-    template <int fe_degree, int n_q_points_1d>
-    static bool
-    run(const unsigned int                     n_components,
-        const EvaluationFlags::EvaluationFlags integration_flag,
-        Number *                               values_dofs,
+        OtherNumber *                          values_dofs,
         FEEvaluationData<dim, Number, false> & fe_eval,
-        const bool                             sum_into_values_array)
+        const bool                             sum_into_values_array = false)
     {
+      // `OtherNumber` is either `const Number` (evaluate()) or `Number`
+      // (integrate())
+      static_assert(
+        std::is_same<Number,
+                     typename std::remove_const<OtherNumber>::type>::value,
+        "Type of Number and of OtherNumber do not match.");
```

Comment:
@kronbichler  I have added a comment/assert that `OtherNumber` is either a const or non-const version of `Number`.